controller/contour/job.go

```diff
@@ -0,0 +1,197 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contour
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// TODO [danehans]: Once https://github.com/projectcontour/contour-operator/pull/30/files
+	// merges, associate contour controller owning label. The job will not get removed without this.
+	//controllerJobLabel = "contour.operator.projectcontour.io/job-contour"
+	jobContainerName = "contour"
+	jobImageName     = "docker.io/projectcontour/contour:main"
+	jobSvcAcctName   = "contour-certgen"
+	jobLabelValue    = "contour-certgen"
+	jobNsEnvVar      = "CONTOUR_NAMESPACE"
+)
+
+// ensureJob ensures that a Job exists for the given contour.
+func (r *Reconciler) ensureJob(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	haveJob, current, err := r.currentJob(ctx, contour)
+	if err != nil {
+		return fmt.Errorf("failed to get job: %w", err)
+	}
+	desired, err := desiredJob(contour)
+	if err != nil {
+		return fmt.Errorf("failed to build job: %w", err)
+	}
+
+	switch {
+	case !haveJob:
+		if err := r.Client.Create(ctx, desired); err != nil {
+			return fmt.Errorf("failed to create job: %w", err)
+		}
+		r.Log.Info("created job", "name", desired.Name)
+		return nil
+	case haveJob:
+		if err := r.updateJob(ctx, current, desired); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// currentJob gets the Job for contour from the api server, returning true
+// and the Job if successful.
+func (r *Reconciler) currentJob(ctx context.Context, contour *operatorv1alpha1.Contour) (bool, *batchv1.Job, error) {
```

Comment:
Rather than trying to update the job in place, what I'd suggest is using GenerateName to name the job (IIRC that can be made to work) and labelling it with an owner ref. That way you can always List() to find the certgen jobs you own, and still generate unique names when you need a new one.

Reply:
> labelling it with an owner ref.

@jpeach the `Contour` that creates the `Job` resides in a different namespace, so it can't be used as an owner. Is the original job the owner?